aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -5,75 +5,247 @@
 
 #include <ATen/core/ivalue.h>
 #include <c10/core/TensorOptions.h>
+#include <ATen/core/Dimname.h>
+
 #include <ATen/core/boxing/KernelFunction.h>
 
-namespace at {
-struct Dimname;
-}
+#include <c10/util/Metaprogramming.h>
 
 namespace c10 {
 namespace impl {
 
+//
+// utils
+//
+
+// is_tensor_ref
+template<class T> struct is_tensor_ref : std::false_type {};
+template <> struct is_tensor_ref<at::Tensor&> : std::true_type {};
+
+// is_tuple_of_tensor_refs
+//
+template<class T, class Enable = void>
+struct is_tuple_of_tensor_refs : std::false_type {};
+
+template<class T>
+struct is_tuple_of_tensor_refs<T, std::enable_if_t<guts::is_instantiation_of<std::tuple, T>::value, void>>
+: guts::typelist::all<is_tensor_ref, guts::typelist::from_tuple_t<T>>
+{};
+
+// has_ivalue_to
+//
+template<class T, class Enable = void>
+struct has_ivalue_to : std::false_type {};
+
+template<class T>
+struct has_ivalue_to<T, guts::void_t<decltype(std::declval<IValue>().to<T>())>>
+: std::true_type
+{};
+
+//
+// boxing predicates
+//
+
+// A boxable arg type is one that IValue has a constructor for.
 // Assume T is decayed
```

Comment:
There is a nasty little asymmetry here, which is that `can_box` assumes the inputs are decayed, but `can_unbox` arguments MUST NOT be decayed (because if they are, then you can't identify `Tensor&` as a return type that cannot be unboxed.)

Reply:
Excellent point! The "assume decayed" always felt fragile to me, but I didn't feel like I had a good enough reason to move it into the body. UNTIL NOW